odl/contrib/tensorflow.py

```diff
@@ -0,0 +1,512 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Utilities for interfacing ODL with Tensorflow."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from builtins import range, str
+from future import standard_library
+standard_library.install_aliases()
+
+from odl.set import LinearSpace, LinearSpaceElement, RealNumbers
+from odl.operator import Operator
+
+import numpy as np
+import uuid
+try:
+    # TODO: Lazy import to improve performance
+    import tensorflow as tf
+    TENSORFLOW_AVAILABLE = True
+except:
+    TENSORFLOW_AVAILABLE = False
+
+
+__all__ = ('as_tensorflow_layer', 'TensorflowSpace', 'TensorflowSpaceOperator')
+
+
+def as_tensorflow_layer(odl_op, default_name='ODLOperator',
+                        differentiable=True):
+    """Convert `Operator` or `Functional` into tensorflow layer.
+
+    Parameters
+    ----------
+    odl_op : `Operator` or `Functional`
+        The operator that should be wrapped to a tensorflow layer.
+    default_name : str
+        Default name for tensorflow layers created.
+    differentiable : boolean
+        True if the layer should be differentiable, in which case  ``odl_op``
+        should implement `Operator.derivative` which in turn implements
+        `Operator.adjoint`, it is properly wrapped in ``tensorflow_layer``, and
+        gradients propagate as expected.
+
+        Otherwise assumes that the derivative is everywhere zero.
+
+    Returns
+    -------
+    tensorflow_layer : callable
+        Callable that, when called with an `tensorflow.Tensor` of shape
+        `(n, *odl_op.domain.shape, 1)` where ``n`` is the number of batches
+        returns another `tensorflow.Tensor`.
+
+        If ``odl_op`` is an `Operator`, the shape of the returned tensor is
+        `(n, *odl_op.range.shape, 1)`.
+
+        Hence for each evaluation, ``odl_op`` is called a total of ``n`` times.
+
+        The `dtype` of the tensor is the same as the respective ODL spaces.
+
+    """
+    if not TENSORFLOW_AVAILABLE:
+        raise ImportError('Tensorflow not installed')
+
+    from tensorflow.python.framework import ops
+
+    def py_func(func, inp, Tout, stateful=True, name=None, grad=None):
+        """Define custom py_func which takes also a grad op as argument.
+
+        We need to overwrite this function since the default tensorflow py_func
+        does not support custom gradients.
```

Comment:
Well you can point to the docs of `py_func` to demonstrate its lack of gradient. Makes it more likely that somebody who follows that link sees that "oh, they added that feature".